File: Analizador Semantico/Analizador/src/Scanner.java
import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Scanner {
    private String source;
    private int linea;
    private List<Token> tokens;
    private Map<String, TokenType> palabras_reservadas;

    public Scanner(String source) {
        this.source = source;
        this.linea = 1;
        this.tokens = new ArrayList<>();
        this.palabras_reservadas = new HashMap<>();

        palabras_reservadas.put("and", TokenType.AND);
        palabras_reservadas.put("class", TokenType.CLASS);
        palabras_reservadas.put("also", TokenType.ALSO);
        palabras_reservadas.put("for", TokenType.FOR);
        palabras_reservadas.put("fun", TokenType.FUN);
        palabras_reservadas.put("if", TokenType.IF);
        palabras_reservadas.put("null", TokenType.NULL);
        palabras_reservadas.put("print", TokenType.PRINT);
        palabras_reservadas.put("return", TokenType.RETURN);
        palabras_reservadas.put("super", TokenType.SUPER);
        palabras_reservadas.put("this", TokenType.THIS);
        palabras_reservadas.put("true", TokenType.TRUE);
        palabras_reservadas.put("var", TokenType.VAR);
        palabras_reservadas.put("while", TokenType.WHILE);
        palabras_reservadas.put("else", TokenType.ELSE);
        palabras_reservadas.put("false", TokenType.FALSE);
        palabras_reservadas.put("or", TokenType.OR);
    }

    public List<Token> ScanTokens() {
        int estado = 0;
        String current = "";

        String[] lines = source.split("\n");
        for (String line : lines) {
            String line1 = clean(line);
            line1 += " ";

            for (char c : line1.toCharArray()) {
                switch (estado) {
                    case 0:
                        if (c == '<') {
                            estado = 1;
                        } else if (c == '=') {
                            estado = 2;
                        } else if (c == '>') {
                            estado = 3;
                        } else if (Character.isDigit(c)) {
                            current += c;
                            estado = 4;
                        } else if (Character.isLetter(c)) {
                            current += c;
                            estado = 5;
                        } else if (c == '/') {
                            estado = 6;
                        } else if (c == '{') {
                            tokens.add(new Token(TokenType.BRACKET_OPEN, "{", null, linea));
                            estado = 0;
                        } else if (c == '}') {
                            tokens.add(new Token(TokenType.BRACKET_CLOSE, "}", null, linea));
                            estado = 0;
                        } else if (c == '(') {
                            tokens.add(new Token(TokenType.PARENT_OPEN, "(", null, linea));
                            estado = 0;
                        } else if (c == ')') {
                            tokens.add(new Token(TokenType.PARENT_CLOSE, ")", null, linea));
                            estado = 0;
                        } else if (c == '+') {
                            tokens.add(new Token(TokenType.ADD, "+", null, linea));
                            estado = 0;
                        } else if (c == '-') {
                            tokens.add(new Token(TokenType.SUB, "-", null, linea));
                            estado = 0;
                        } else if (c == '*') {
                            tokens.add(new Token(TokenType.MULT, "*", null, linea));
                            estado = 0;
                        } else if (c == '!') {
                            estado = 8;
                        } else if (c == '"') {
                            current += c;
                            estado = 9;
                        } else if (c == ';') {
                            tokens.add(new Token(TokenType.SEMICOLON, ";", null, linea));
                            estado = 0;
                        } else if (c == ',') {
                            tokens.add(new Token(TokenType.COMMA, ",", null, linea));
                            estado = 0;
                        }
                        break;
                    case 1:
                        if (c == '=') {
                            tokens.add(new Token(TokenType.LESS_EQUAL, "<=", null, linea));
                            estado = 0;
                        } else {
                            tokens.add(new Token(TokenType.LESS_THAN, "<", null, linea));
                            estado = 0;
                        }
                        break;
                    case 2:
                        if (c == '=') {
                            tokens.add(new Token(TokenType.EQUAL, "==", null, linea));
                            estado = 0;
                        } else {
                            tokens.add(new Token(TokenType.ASIGNATION, "=", null, linea));
                            estado = 0;
                        }
                        break;
                    case 3:
                        if (c == '=') {
                            tokens.add(new Token(TokenType.GREAT_EQUAL, ">=", null, linea));
                            estado = 0;
                        } else {
                            tokens.add(new Token(TokenType.GREAT, ">", null, linea));
                            estado = 0;
                        }
                        break;
                    case 4:
                        if (Character.isDigit(c) || c == '.') {
                            current += c;
                        } else {
                            tokens.add(new Token(TokenType.NUMBER, current, Float.parseFloat(current), linea));
                            current = "";
                            estado = 0;
                        }
                        break;
                    case 5:
                        if (Character.isDigit(c) || Character.isLetter(c)) {
                            current += c;
                        } else {
                            if (palabras_reservadas.containsKey(current)) {
                                if (current.toLowerCase().equals("true")) {
                                    tokens.add(new Token(TokenType.TRUE, current, true, linea));
                                    current = "";
                                    estado = 0;
                                } else if (current.toLowerCase().equals("false")) {
                                    tokens.add(new Token(TokenType.FALSE, current, false, linea));
                                    current = "";
                                    estado = 0;
                                } else {
                                    tokens.add(new Token(palabras_reservadas.get(current), current, null, linea));
                                    current = "";
                                    estado = 0;
                                }
                            } else {
                                tokens.add(new Token(TokenType.IDENTIFIER, current, null, linea));
                                current = "";
                                estado = 0;
                            }
                        }
                        break;
                    case 6:
                        if (c == '/') {
                            estado = 7;
                        } else if (c == '*') {
                            estado = 11;
                        } else {
                            tokens.add(new Token(TokenType.DIAG, "/", null, linea));
                            estado = 0;
                        }
                        break;
                    case 7:
                        if (c == '\n') {
                            estado = 0;
                        } else {
                            estado = 7;
                        }
                        break;
                    case 8:
                        if (c == '=') {
                            tokens.add(new Token(TokenType.DIFERENT, "!=", null, linea));
                            estado = 0;
                        } else {
                            tokens.add(new Token(TokenType.NEGATION, "!", null, linea));
                            estado = 0;
                        }
                        break;
                    case 9:
                        if (c == '"') {
                            current += c;
                            tokens.add(new Token(TokenType.STRING, current, current.substring(1, current.length() - 1), linea));
                            current = "";
                            estado = 0;
                        } else {
                            current += c;
                        }
                        break;
                    case 11:
                        if (c == '*') {
                            estado = 12;
                        }
                        break;
                    case 12:
                        if (c == '/') {
                            estado = 0;
                        } else {
                            estado = 11;
                        }
                        break;
                }
            }
            linea++;
        }
        tokens.add(new Token(TokenType.EOF, null, null, linea - 1));
        return tokens;
    }

    private String clean(String cadena) {
        String[] simbolos = { "(", ")", "{", "}", "=", "<", ">", "!", "+", "-", ";", "*", "/" };
        String clean_str = "";
        String pattern = "\\/\\/.*|\\/\\*[\\s\\S]*?\\*\\/|([A-Za-z_][A-Za-z0-9_]*|\\d+(?:\\.\\d+)?|\\S)";
        Pattern r = Pattern.compile(pattern);
        Matcher m = r.matcher(cadena);
        List<String> result = new ArrayList<>();
        while (m.find()) {
            result.add(m.group());
        }
        clean_str = String.join(" ", result);
        clean_str = clean_str.replace("/ *", "/*");
        clean_str = clean_str.replace("* /", "*/");
        clean_str = clean_str.replace("> =", ">=");
        clean_str = clean_str.replace("< =", "<=");
        clean_str = clean_str.replace("= =", "==");
        clean_str = clean_str.replace("! =", "!=");

        return clean_str + "\n";
    }
}

File: Analizador Semantico/Scanner.py
import re
from Tokentype import TokenType #Importar la clase TokenType del módulo tipo_token
from Token import Token #Importar la clase Token del módulo token
import string  #Importar el módulo string

class Scanner: #Clase Scanner
    def __init__(self,source) -> None: #Definir el método constructor de la clase
        self.source = source
        self.__linea = 1  #Asignar el atributo linea a la instancia de la clase, con valor inicial de 1
        self.tokens = []  #Asignar el atributo tokens a la instancia de la clase, con una lista vacía como valor inicial
        self.palabras_reservadas = { # Diccionario con las palabras reservadas y su tipo de token correspondiente
            # "palabra" - "tipo token" 
            'and': TokenType.AND,
            'class': TokenType.CLASS,
            'also': TokenType.ALSO,
            'for': TokenType.FOR,
            'fun' : TokenType.FUN,
            'if' : TokenType.IF,
            'null' : TokenType.NULL,
            'print' : TokenType.PRINT,
            'return' : TokenType.RETURN,
            'super' : TokenType.SUPER,
            'this' : TokenType.THIS,
            'true' : TokenType.TRUE,
            'var' : TokenType.VAR,
            'while' : TokenType.WHILE,
            'else': TokenType.ELSE,
            'true': TokenType.TRUE,
            'false': TokenType.FALSE,
            'or': TokenType.OR
        } 

    def ScanTokens(self): # -> list[TokenType]:
        ##Aqui va el scanner


        self.estado = 0

        for line in self.source:
            current = ""
            line1 = self.clean(line)
            line1 +=" "
             # Ciclo for que recorre cada caracter en la línea ya limpia
            for char in line1:
                match self.estado: # Inicia un bloque de comparaciones de acuerdo al valor de la variable estado
                    case 0:
                        if char == "<": # Dependiento del caracter actual el en este caso es "<", se cambia el valor de estado a 1
                            self.estado = 1
                        elif char == "=":
                            self.estado = 2
                        elif char == ">":
                            self.estado = 3
                        elif char.isdigit(): # Si el caracter actual es un dígito, se agrega a la variable current y se cambia el valor de estado a 4
                           current += char
                           self.estado = 4
                        elif char.isalpha():
                            current += char
                            self.estado = 5
                        elif char == "/":
                            self.estado = 6
                        elif char == "{": # Si el caracter actual es "{", se agrega un token de tipo TokenType.ALLAVE a la lista tokens, y se cambia el valor de estado a 0
                            self.tokens.append(Token(TokenType.BRACKET_OPEN,"{",None,self.__linea))
                            self.estado = 0
                        elif char == "}":
                            self.tokens.append(Token(TokenType.BRACKET_CLOSE,"}",None,self.__linea))
                            self.estado = 0
                        elif char == "(":
                            self.tokens.append(Token(TokenType.PARENT_OPEN,"(",None,self.__linea))
                            self.estado = 0
                        elif char == ")":
                            self.tokens.append(Token(TokenType.PARENT_CLOSE,")",None,self.__linea))
                            self.estado = 0
                        elif char == "+":
                            self.tokens.append(Token(TokenType.ADD,"+",None,self.__linea))
                            self.estado = 0
                        elif char == "-":
                            self.tokens.append(Token(TokenType.SUB,"-",None,self.__linea))
                            self.estado = 0
                        elif char == "*":
                            self.tokens.append(Token(TokenType.MULT,"*",None,self.__linea))
                            self.estado = 0
                        elif char == "!":
                            self.estado = 8
                        elif char == '"':
                            current += char
                            self.estado = 9
                        elif char == ";":
                            self.tokens.append(Token(TokenType.SEMICOLON,";",None,self.__linea))
                            self.estado = 0
                        elif char == ",":
                            self.tokens.append(Token(TokenType.COMMA,",",None,self.__linea))
                            self.estado = 0
                        else:# Si el carácter actual no es ninguno de los anteriores: No hacemos nada y pasamos al siguiente carácter.
                            pass
                    case 1: # Si el estado actual es 1:
                        if char == "=":
                            self.tokens.append(Token(TokenType.LESS_EQUAL,"<=",None,self.__linea)) # Añadimos un nuevo objeto Token a la lista de tokens.
                            self.estado = 0 # Cambiamos el estado del scanner a 0.
                        else:
                            self.tokens.append(Token(TokenType.LESS_THAN,"<",None,self.__linea))
                            self.estado = 0
                    case 2:
                        if char == "=":
                            self.tokens.append(Token(TokenType.EQUAL,"==",None,self.__linea))
                            self.estado = 0
                        else:
                            self.tokens.append(Token(TokenType.ASIGNATION,"=",None,self.__linea))
                            self.estado = 0
                    case 3:
                        if char == "=":
                            self.tokens.append(Token(TokenType.GREAT_EQUAL,">=",None,self.__linea))
                            self.estado = 0
                        else:
                            self.tokens.append(Token(TokenType.GREAT,">",None,self.__linea))
                            self.estado=0
                        pass
                    case 4:
                        if char.isdigit() or char == ".":
                            current += char
                        else:
                            self.tokens.append(Token(TokenType.NUMBER,current,float(current),self.__linea))
                            current = ""
                            self.estado = 0
                    case 5:
                        if char.isdigit() or char.isalpha():
                            current += char
                        else:
                            if current in self.palabras_reservadas:
                                #print(f"{current}")
                                if current.lower() == "true":
                                    self.tokens.append(Token(TokenType.TRUE,current,True,self.__linea))
                                    current = ""
                                    self.estado = 0
                                elif current.lower() == "false":
                                    self.tokens.append(Token(TokenType.FALSE,current,False,self.__linea))
                                    current = ""
                                    self.estado = 0
                                else:
                                    self.tokens.append(Token(self.palabras_reservadas[current],current,None,self.__linea))
                                    current = ""
                                    self.estado = 0
                            else:
                                self.tokens.append(Token(TokenType.IDENTIFIER,current,None,self.__linea))
                                current = ""
                                self.estado = 0
                    case 6:
                        if char == "/":
                            self.estado = 7
                        elif char == "*":
                            self.estado = 11
                        else:
                            self.tokens.append(Token(TokenType.DIAG,"/",None,self.__linea))
                            self.estado = 0
                    case 7:
                        if char == "\n":
                            self.estado = 0
                        else:
                            self.estado = 7
                    case 8:
                        if char == "=":
                            self.tokens.append(Token(TokenType.DIFERENT,"!=",None,self.__linea))
                            self.estado = 0
                        else:
                            self.tokens.append(Token(TokenType.NEGATION,"!",None,self.__linea))
                            self.estado = 0
                    case 9:
                        if char == '"':
                            current += char
                            self.tokens.append(Token(TokenType.STRING,current,current[2:-2],self.__linea))
                            current = ""
                            self.estado = 0
                        else:
                            current += char
                    case 11:
                        if char == "*":
                            self.estado = 12
                    case 12:
                        if char == "/":
                            self.estado = 0
                        else:
                            self.estado = 11

            self.__linea += 1 # Despues de leer la primera linea se incrementa
                        
        self.tokens.append(Token(TokenType.EOF,None,None,self.__linea-1)) # Se termina el archivo y se agrega el token EOF
        return self.tokens # Devuelve la lista de tokens generada

    def clean(self,cadena): # Define una función 'clean' que toma una cadena como entrada y devuelve una cadena limpia
        simbolos = ['(', ')' ,'{' ,'}', '=', '<', '>', '!', '+', '-', ';', '*', '/']   # Define una lista de Símbolos
        clean_str = ''

        pattern = r'\/\/.*|\/\*[\s\S]*?\*\/|([A-Za-z_][A-Za-z0-9_]*|\d+(?:\.\d+)?|\S)'
        result = re.findall(pattern, cadena)    
        
        clean_str = " ".join(result)

        

        clean_str=clean_str.replace("/ *","/*")
        clean_str=clean_str.replace("* /","*/")
        clean_str=clean_str.replace("> =",">=")
        clean_str=clean_str.replace("< =","<=")
        clean_str=clean_str.replace("= =","==")
        clean_str=clean_str.replace("! =","!=")

        #print(clean_str)

        return f"{clean_str}\n"

        
